### DISCUSSION TITLE
Bring up Samsung proprietary VK driver on proot linux

### DISCUSSION THREAD

**Discussion by mshzhb** [termux/termux-packages#22451] [Q&A]
I'm trying to get Samsung's proprietary Vulkan driver (VK UMD) working on Termux-X11 proot linux. I heard that a special patch is needed for Termux-X11 since it differs from standard X11. The Samsung Vulkan driver has native X server & DRI3 support, but it seems like Termux-X11 might require modifications for compatibility.

Does anyone have experience or guidance on patching Termux-X11 for this setup? Any resources or help would be greatly appreciated. (my email: [tong.liu@samsung.com](mailto:tong.liu@samsung.com))

Thanks in advance!

**Comment by twaik** [comment]
Currently it is possible to use proprietary drivers with termux-x11 with xMeM's vulkan-wsi-layer [mesa-vulkan-icd-wrapper](https://github.com/xMeM/termux-packages/commit/854358442b88d7dcf6b1694f4047df48d9b5c925). I do not add regular DRM/KMS to termux x11 because there is no public Android API for that and even if I will use some non-public APIs it will break rendering on other devices.

**Reply by twaik** [reply]
Oh, the question is about proot. Probably you can not use any of them here...

**Reply by mshzhb** [reply]
Thanks a lot for the immediate response! I'm intending to bring it up on proot linux. Probably termux + glibc environment later. I'm new to this environment. 

Currently, the vulkaninfo is working. However, vkcube fails in window creation.

**Reply by twaik** [reply]
Start termux-x11 with `-disable-dri3` flag.

**Reply by mshzhb** [reply]
Hi @twaik, is it possible to use wayland in termux instead of x server?

**Reply by mshzhb** [reply]
offscreen rendering and compute are working now

**Reply by twaik** [reply]
With labwc (from PR) or weston.

**Comment by mshzhb** [comment]
offscreen rendering is working.
However, having some issues with **xcb_dri3_open_reply**
Here is the error code: DRI3 related

error_code = 0x8  (BadMatch)
major_code = 0x93 (dri3 related)
minor_code = 0x1

**Reply by twaik** [reply]
Because termux-x11's dri3 is not a full implementation your driver expects. You can send or ahardwarebuffer or some mmappable fd.

**Reply by mshzhb** [reply]
Thanks a lot sir! That's the answer I'm looking for! Since turnip can work on termux-x11, there should be a way for samsung(AMD) driver to work

**Reply by mshzhb** [reply]
Do you have any more information on this topic? Who makes turnip working on termux?

**Reply by twaik** [reply]
https://github.com/termux/termux-packages/blob/master/packages/mesa-vulkan-icd-freedreno-dri3/wsi-termux-x11.patch

**Comment by mshzhb** [comment]
Hi @twaik & @xMeM,

For on-screen rendering, our driver is already functional on proot Linux with CPU blit, but I’m currently working on enabling DRI3. Could you clarify whether, for Turnip to work with Termux-X11, I only need to apply [wsi-termux-x11.patch](https://github.com/termux/termux-packages/blob/master/packages/mesa-vulkan-icd-freedreno-dri3/wsi-termux-x11.patch), or is the [tu_kgsl_export_dmabuf.patch](https://github.com/termux/termux-packages/blob/master/packages/mesa-vulkan-icd-freedreno-dri3/tu_kgsl_export_dmabuf.patch) required as well?

Currently, xcb_dri3_pixmap_from_buffers_checked(..., 1274, &fd) is returning error_code = 11.

Thanks,
Tong
![20241216_204659](https://github.com/user-attachments/assets/f9c9af40-392d-48d4-b21e-1362f3a34a2c)

**Reply by xMeM** [reply]
mesa version higher than `24.2.0` only needs to apply `wsi-termux-x11.patch`

**Reply by mshzhb** [reply]
Hi @xMeM, could you please briefly explain why [tu_kgsl_export_dmabuf.patch](https://github.com/termux/termux-packages/blob/master/packages/mesa-vulkan-icd-freedreno-dri3/tu_kgsl_export_dmabuf.patch) is nolonger needed? Our drivers (proprietary) fails at xcb_dri3_pixmap_from_buffers_checked(..., 1274, &fd), error_code = 11 (bad allocation). I suspect the fd is not mmappable. Thanks!

**Reply by xMeM** [reply]
`tu_kgsl_export_dmabuf.patch` has already been implemented upstream.

**Reply by xMeM** [reply]
You can allocate a mappable dmabuf fd from dma_heap or ion heap, then import the dmabuf fd into the swapchain image memory with `VK_EXT_external_memory_dma_buf`

**Reply by twaik** [reply]
Is it possible to do the same thing with EGLImage? I mean allocate some dma or ion heap and import it to egl?

**Reply by xMeM** [reply]
Requires driver support for `EGL_EXT_image_dma_buf_import.`

**Reply by mshzhb** [reply]
<img width="601" alt="image" src="https://github.com/user-attachments/assets/7cbd6859-1ad3-4088-8e13-cf546b421d15" />

Thanks a lot @twaik & @xMeM. DRI3 is working now. The last task I need to do is fixing xshmfence according to [[wsi-termux-x11.patch](https://github.com/termux/termux-packages/blob/master/packages/mesa-vulkan-icd-freedreno-dri3/wsi-termux-x11.patch)], isn't it?

Btw, in this setup can I use optimal tiling for the framebuffer? Currently only linear works.

Currently I'm unclear why this turnip branch (https://gitlab.freedesktop.org/Pipetto-crypto/mesa/-/commits/winlator_wsi-termux-x11?search=termux) can work while still keep the xshmfence

**Reply by xMeM** [reply]
termux-x11 uses `pixman` (cpu based acceleration) which can only handle linear tiling. When the framebuffer uses optimal tiling, it needs to be copied to a linear buffer.

This branch (https://gitlab.freedesktop.org/Pipetto-crypto/mesa/-/commits/winlator_wsi-termux-x11?search=termux) use `xcb_sync_fence` instead of `xshmfence,` it still removes `xshmfence.`

**Reply by twaik** [reply]
XCB fence is a valid option even if it will be a bit slower.

**Reply by mshzhb** [reply]
Thanks for the guidance! With xcb_sync_fence, our umd is fully working now!
